bin/utest/mmap.c

```diff
@@ -93,3 +96,38 @@ int test_mmap(void) {
   munmap_bad();
   return 0;
 }
+
+static volatile int sigsegv_handled = 0;
+static jmp_buf return_to;
+static void sigsegv_handler(int signo) {
+  printf("sigsegv handled!\n");
+  sigsegv_handled++;
+  longjmp(return_to, 5);
+}
+int test_mmap_readonly(void) {
+  size_t pgsz = getpagesize();
+  signal(SIGSEGV, sigsegv_handler);
+  void *addr = mmap(NULL, pgsz * 8, PROT_READ, MAP_ANON | MAP_PRIVATE, -1, 0);
+  assert(addr != MAP_FAILED);
+  printf("mmap returned pointer: %p\n", addr);
+
+  /* Ensure mapped area is cleared. */
+  assert(*(char *)(addr + 100) == 0);
+  assert(*(char *)(addr + 1000) == 0);
+
+  sigsegv_handled = 0;
+
+  if (setjmp(return_to) == 0) {
+    printf("Try to write to readonly memory\n");
+    /* Try to write to readonly memory. It should raise SIGSEGV */
+    *(char *)addr = '9';
+  }
+
+  assert(sigsegv_handled == 1);
+  assert(*(char *)addr == 0);
+
+  /* restore original behavior */
+  signal(SIGSEGV, SIG_DFL);
+
+  return 0;
+}
```

Comment:
I think that you don't need non-local jumps here.

Reply:
I need it 'cause I have to handle sigsegv. Also I replaced setjmp and longjmp with sigsetjmp and siglongjmp